### PR TITLE
Association

### PR DESCRIPTION
**Pull request by TaiwoEnoch**
## I create the PR to implement setting up of association ⬇️ 

- [ ] Create a class Classroom with the following:

- @label instance variable, should be initialized in the constructor.
- Setter and getter for @label (remember about attr_accessor).

- [ ] Create the has-many/belongs-to relationship between Classroom and Student. The following should be implemented:

- Create the has-many side (a classroom has many students).
- Create the belongs-to side (a student belongs to a classroom).
- Make sure that when adding a student to a classroom it also sets the classroom for the student.
- Make sure that when setting the classroom for a student it also adds it to the classrooms' students.

- [ ] Create a class Book with the following:

- @title and @author instance variables, should be initialized in the constructor.
- Setters and getters for instance variables (remember about attr_accessor).

- [ ] Create a class Rental with the following:

- @date instance variable, should be initialized in the constructor.
- Setter and getter for @date (remember about attr_accessor).

- [ ] Create the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. The following should be implemented:

- Create the has-many side of Book and Rental (a book has many rentals).
- Create the belongs-to side of Rental and Book (a rental belongs to a book).
- Create the has-many side of Person and Rental (a person has many rentals).
- Create the belongs-to side of Rental and Person (a rental belongs to a person).
- Modify the constructor of Rental so Book and Person are set in it.

### CODE REVIEW

File: student.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+  def initialize(id, age, parent_permission: true, name: 'Unknown')
+    super(id, age, parent_permission: parent_permission, name: name)
+    @classroom = nil
+  end
+
+  def add_to_classroom(classroom)
+    @classroom = classroom
+    classroom.add_student(self)
+  end
```
Comment (mikethreels) [inline student.rb]: - [ ] the method that serves as the 'belongs-to' side (a student belongs to a classroom) should look a little different. Have a look at [this Link](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#both-ways-relationships) to the article with examples.

File: book.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+class Book
+  attr_accessor :title, :author
+  attr_reader :rentals
```
Comment (mikethreels) [inline book.rb]: - [ ] in order for this to work Getters and Setters for @rentals should be present

File: book.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals << rental
+    rental.book = self
+  end
```
Comment (mikethreels) [inline book.rb]: - [ ] A method for adding rentals should look a little different. please have a look at the example in a previous lesson under [this Link](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship)

File: rental.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+class Rental
+  attr_accessor :date
+  attr_reader :person, :book
```
Comment (mikethreels) [inline rental.rb]: - [ ] in order for this to work Getters and Setters for @book and @person instance variables should be present

File: person.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
+require_relative 'nameable'
+
+class Person < Nameable
+  attr_reader :id, :age
+  attr_accessor :name, :rentals
```
Comment (mikethreels) [inline person.rb]: - [ ] In order for this to work the Getters and Setters for @age, @name & @rentals instance variables should be present

File: person.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  end
+
+  def can_use_services?
+    of_age? || @parent_permission
+  end
+
+  def add_rental(rental)
+    @rentals << rental
+    rental.person = self
+  end
```
Comment (mikethreels) [inline person.rb]: - [ ] A method for adding rentals should look a little different. please have a look at the example in a previous lesson under [this Link](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship)

**Review by mikethreels — COMMENTED** [review]
Hi Enoch, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- Good job starting with the associations.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: book.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+class Book
+  attr_accessor :title, :author
```
Comment (OyePriscilla) [inline book.rb]: - [ ] As raised earlier, kindly add @rentals to be part of the attribute accessor. 👍 
```suggestion
class Book
  attr_accessor :title, :author, :rentals
```

File: rental.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+class Rental
+  attr_accessor :date
```
Comment (OyePriscilla) [inline rental.rb]: - [ ] @book and @person suppose to have a getter and setter method. Kindly implement this by using the attribute accessor. 👍 
```suggestion
class Rental
  attr_accessor :date, :book, :person
```

**Review by OyePriscilla — CHANGES_REQUESTED** [review]
Hi @TaiwoEnoch 👏 👏, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

so-close-so-close-seth-meyers

 ## Highlights
Implemented a method that serves as the 'belongs-to' relationship in ruby ✔️ 
Set up associations between classes and objects ✔️ 

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: rental.rb (hunk 8 above)
Comment (TaiwoEnoch) [reply rental.rb]: Thanks so much for the clarity @OyePriscilla

**Review by binhussen — APPROVED** [review]
Hi @TaiwoEnoch , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉

## Highlights
- Fixed Reviewer feedback✔️
- Set up associations ✔️
- Used Git flow ✔️
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**